### ISSUE TITLE
feat: AI-powered genre tagging and music embeddings

### ISSUE THREAD

**Issue by zzstoatzz** [zzstoatzz/plyr.fm#847]
## context

inspired by [this thread](https://bsky.app/profile/piss.beauty/post/3lhcn5ibg5c2g) - the idea is to use ML models to automatically suggest genre tags for uploaded tracks.

> "an actual good use of models in music would be genre labeling. 'computer is this deep tech or just deep house. give me your top 2-3 genre tags'" — @piss.beauty

key insight from the thread: embeddings for similarity search, with semantic labels as a "shorthand on top" - do embeddings first, then reduce to "this vector is closest to deep house"

### 1. Replicate effnet-discogs API (recommended starting point)

[replicate.com/mtg/effnet-discogs](https://replicate.com/mtg/effnet-discogs)

- **cost**: ~$0.00019/run (~2 seconds on CPU)
- **output**: 400 Discogs styles (deep house, techno, acid, etc.) with confidence scores
- **input**: audio file URL or YouTube URL
- **format**: JSON or visualization

pros:
- zero infra, just API calls
- discogs taxonomy is industry-standard for electronic music
- cheap enough for all uploads

cons:
- external dependency
- CC BY-NC-SA license on underlying model (may need commercial license)

### 2. self-hosted Essentia + discogs-effnet

[essentia.upf.edu/models.html](https://essentia.upf.edu/models.html)

- same model as Replicate but self-hosted
- could integrate into existing transcoder service (Rust)
- 1280-dimensional embeddings available for similarity search

pros:
- no per-request cost
- full control, can extract embeddings for similarity
- already have transcoder infra

cons:
- more complex deployment
- CC BY-NC-SA 4.0 (non-commercial), need license for production
- TensorFlow/ONNX runtime overhead

### 3. Cyanite.ai (enterprise option)

[cyanite.ai](https://cyanite.ai/)

- 35M+ songs tagged, GraphQL API
- genre, mood, instruments, lyric theme, tempo
- EU-based, GDPR compliant

pros:
- most comprehensive tagging
- production-ready

cons:
- enterprise pricing (likely expensive)
- overkill for MVP

### 4. cosine.club approach

[cosine.club](https://cosine.club/) - reference implementation

- uses discogs-effnet embeddings
- 1.1M+ tracks indexed
- cosine similarity for "find similar"

this is the dream UX: type a track, get similar tracks. could be a future feature.

### phase 1: genre suggestions on upload
1. after transcoding completes, call effnet-discogs API (Replicate)
2. store top-N style predictions in track record
3. show suggestions to user during upload, let them accept/edit
4. store final tags in ATProto track record

### phase 2: embeddings for similarity (future)
1. store 1280-dim embedding vector per track
2. use pgvector or similar for similarity search
3. "find similar tracks" feature
4. playlist generation based on sonic similarity

## data model considerations

where do tags live?
- ATProto track record (portable, decentralized)
- local db only (faster queries, but not portable)
- both (source of truth in ATProto, cache locally)

suggested track record extension:
```json
{
  "genres": ["Deep House", "Tech House"],
  "genreConfidence": [0.87, 0.65],
  "embedding": "base64-encoded-vector"  // or separate record
}
```

## open questions

1. **user agency**: auto-tag vs suggest-and-confirm?
2. **embeddings**: worth storing for similarity? adds complexity
3. **taxonomy**: Discogs styles vs custom? interop with other apps?
4. **rate limits**: tag on upload only, or batch-tag existing catalog?

## references

- [Essentia models](https://essentia.upf.edu/models.html) - discogs-effnet and other pre-trained models
- [cosine.club](https://cosine.club/) - music similarity search using same approach
- [Cyanite.ai](https://cyanite.ai/) - enterprise music tagging API
- [Music Genome Project](https://en.wikipedia.org/wiki/Music_Genome_Project) - Pandora's approach from 25 years ago
- [MusicBrainz](https://musicbrainz.org/) - open music metadata database
